### PR TITLE
Add `maybe_code_feedback()` to default `fail()` message

### LINKED ISSUE #171: `library(gradethis)` should auto call `gradethis_setup()` if knitr is in progress
Related issue: https://github.com/rstudio/learnr/issues/448
Related code: https://github.com/rstudio/learnr/blob/8ecaab8244567e7e38e610d48a48b35af52b545d/R/initialize.R#L14

This would help make sure _something_ is available for checking.

### LINKED ISSUE #172: add `feedback` option to `grade_this*()`
Similar to https://github.com/rstudio-education/gradethis/issues/136

Thinking `gradethis` should listen to an option set when making the default message

### PR DESCRIPTION
**Pull request by schloerke**
Fixes #171
Fixes #172

----------------------

Allows authors to do...

* No feedback
```r
grade_this({
  fail()
}, fail_code_feedback = FALSE)
# "Incorrect."
```
* With feedback
```r
grade_this({
  fail()
}, fail_code_feedback = TRUE)
# "Incorrect. <Reason why your code is wrong>"

# Same as default behavior...
grade_this({
  fail()
})
# "Incorrect. <Reason why your code is wrong>"
```
* Custom Message (no feedback)
```r
grade_this({
  fail("Custom Message")
})
# "Custom Message"
```
* Custom Message (w/ feedback)
```r
grade_this({
  fail("Custom Message.{maybe_code_feedback()}")
})
# "Custom Message. <Reason why your code is wrong>"
```

`maybe_code_feedback()` requires `code_feedback()` to succeed without an error.  If an error occurs, then `""` is returned for easy `glue()`ing.

### CODE REVIEW

File: R/zzz.R [hunk 1]
```diff
@@ -20,10 +30,31 @@ gradethis_default_options <- list(
...
+      names(gradethis_learnr_knitr_opts),
+      gradethis_learnr_knitr_opts,
+      f = function(key, val) {
+        # only set if the current value is NULL
+        if (is.null(knitr::opts_chunk$get(key))) {
+          knitr::opts_chunk$set(key, val)
+        }
+      }
+    )
+  }
```
Comment (schloerke) [inline R/zzz.R:57]: Fixes #171 
cc @gvwilson

File: R/zzz.R [hunk 2]
```diff
@@ -20,10 +30,31 @@ gradethis_default_options <- list(
...
+)
+
 .onLoad <- function(libname, pkgname) {
+  # initialize gradethis global options
   op <- options()
   toset <- !(names(gradethis_default_options) %in% names(op))
   if (any(toset)) options(gradethis_default_options[toset])
 
+  # initialize tutorial options
+  if (requireNamespace("knitr")) {
```
Comment (cpsievert) [inline R/zzz.R:46]:
```suggestion
  if ("knitr" %in% loadedNamespaces()) {
```

File: R/zzz.R [hunk 3]
```diff
@@ -20,10 +30,31 @@ gradethis_default_options <- list(
   gradethis_glue_incorrect_test = "{ .num_correct }/{ .num_total } correct! { random_encourage() }"
 )
 
+gradethis_learnr_knitr_opts <- list(
```
Comment (cpsievert) [inline R/zzz.R:33]:
```suggestion
# These values should mimic the values set in gradethis_setup()
gradethis_learnr_knitr_opts <- list(
```

Comment (cpsievert) [reply R/zzz.R:33]: Or add a test

File: R/grade_this_code.R [hunk 4]
```diff
@@ -65,13 +67,16 @@
...
+  ...,
+  fail_code_feedback = getOption("gradethis.code.feedback", TRUE),
   allow_partial_matching = getOption("gradethis.code.partial_matching", TRUE)
 ) {
 
   # MUST wrap calling function to be able to shim in `.correct`/`.incorrect`
   function(checking_env) {
     checking_env[[".__correct"]] <- correct
     checking_env[[".__incorrect"]] <- incorrect
+    checking_env[[".__fail_code_feedback"]] <- isTRUE(fail_code_feedback)
```
Comment (cpsievert) [inline R/grade_this_code.R:79]: What's going on here?

File: R/grade_this_code.R [hunk 5]
```diff
@@ -65,13 +67,16 @@
 grade_this_code <- function(
   correct = getOption("gradethis.code.correct", getOption("gradethis.pass", "Correct!")),
   incorrect = getOption("gradethis.code.incorrect", getOption("gradethis.fail", "Incorrect")),
+  ...,
+  fail_code_feedback = getOption("gradethis.code.feedback", TRUE),
```
Comment (cpsievert) [inline R/grade_this_code.R:71]:
```suggestion
```

File: R/grade_this_code.R [hunk 6]
```diff
@@ -65,13 +67,16 @@
...
 ) {
 
   # MUST wrap calling function to be able to shim in `.correct`/`.incorrect`
   function(checking_env) {
     checking_env[[".__correct"]] <- correct
     checking_env[[".__incorrect"]] <- incorrect
+    checking_env[[".__fail_code_feedback"]] <- isTRUE(fail_code_feedback)
 
     grade_this({
       # create variable `.message` for glue to find
```
Comment (cpsievert) [inline R/grade_this_code.R:82]: Whats going on on 83?

File: R/grade_this_code.R [hunk 7]
```diff
@@ -136,40 +143,83 @@ grade_this_code <- function(
...
+
+should_display_code_feedback <- function() {
+  isTRUE(getOption("gradethis.code.feedback.set", FALSE))
+}
+with_code_feedback <- function(val, expr) {
+  with_options(
+    list("gradethis.code.feedback.set" = val),
+    expr
+  )
+}
```
Comment (cpsievert) [inline R/grade_this_code.R:188]:
```suggestion
should_display_code_feedback <- function() {
  isTRUE(getOption("gradethis.code.feedback", FALSE))
}
with_code_feedback <- function(val, expr) {
  with_options(
    list("gradethis.code.feedback" = val),
    expr
  )
}
```

**Review by cpsievert — APPROVED** [review]

**Comment by schloerke** [comment]
- [ ] Merge `maybe_code_feedback()` functionality into `code_feedback()`
- [ ] Remove `maybe_code_feedback()`

**Comment by schloerke** [comment]
Did not pursue dropping `maybe_code_feedback()` as it has a different goal of a _safe_ `code_feedback()` function call. Removing the errors thrown or different behavior of code_feedback seemed like an argument explosion that didn't seem right.

Keeping separate functions for now.

**Review by cpsievert — APPROVED** [review]

**Pull request merged**